Directly operate with CMutableTransaction in SignSignature

martinus opened the pull request:
Refactored `TransactionSignatureCreator` into a templated `GenericTransactionSignatureCreator` that works with both `CMutableTransaction` and `CTransaction`.

The advantage is that now in `SignSignature`, the `MutableTransactionSignatureCreator` can now operate directly with the `CMutableTransaction` without the need to copy the data into a `CTransaction`.

Running all unit tests brings a very noticable speedup on my machine:

    48.4 sec before this change
    36.4 sec with this change
    --------
    12.0 seconds saved

running only `--run_test=transaction_tests/test_big_witness_transaction`:

    16.7 sec before this change
     5.9 sec with this change
    --------
    10.8 seconds saved

This relates to my first attempt with the const_cast hack #13202, and to the slow unit test issue #10026.

Also see #13050 which modifies the tests but not the production code (like this PR) to get a speedup.

martinus committed:
Generic TransactionSignatureCreator works with both CTransaction and CMutableTransaction

Templated version so that no copying of CMutableTransaction into a CTransaction is
necessary. This speeds up the test case transaction_tests/test_big_witness_transaction
from 7.9 seconds to 3.1 seconds on my machine.

promag commented:
Nice improvement, concept ACK.

Empact commented on src/script/interpreter.h:
```diff
@@ -124,7 +124,8 @@ struct PrecomputedTransactionData
 
-    explicit PrecomputedTransactionData(const CTransaction& tx);
+    template<class T>
+    PrecomputedTransactionData(const T& tx);
```
Why drop `explicit` here? I don't see changes to the call sites.

martinus replied:
you are right, I probably should not drop that.

promag commented on src/script/interpreter.h:
```diff
@@ -160,10 +162,11 @@ class BaseSignatureChecker
 
-class TransactionSignatureChecker : public BaseSignatureChecker
+template<class T>
+class GenericTransactionSignatureChecker : public BaseSignatureChecker
```
nit, drop `Generic` prefix?

martinus replied:
I've added it so that the names `TransactionSignatureChecker` and `MutableTransactionSignatureChecker` are still valid as used before

promag replied:
Right.

promag commented:
utACK after squash.

practicalswift commented:
Very nice!

Concept ACK

sipa commented on src/script/sign.h:
```diff
@@ -36,26 +36,23 @@ class BaseSignatureCreator {
-/** A signature creator for transactions. */
-class TransactionSignatureCreator : public BaseSignatureCreator {
-    const CTransaction* txTo;
+template <class T>
```
You don't actually need a template here; all existing uses of `TransactionSignatureCreator` become more efficient to rewrite using `MutableTransactionSignatureCreator`:
```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 2a2f8b5b20..c2b1915b11 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -2608,7 +2608,6 @@ bool CWallet::SignTransaction(CMutableTransaction &tx)
     AssertLockHeld(cs_wallet); // mapWallet
 
     // sign the new tx
-    CTransaction txNewConst(tx);
     int nIn = 0;
     for (const auto& input : tx.vin) {
         std::map<uint256, CWalletTx>::const_iterator mi = mapWallet.find(input.prevout.hash);
@@ -2618,7 +2617,7 @@ bool CWallet::SignTransaction(CMutableTransaction &tx)
         const CScript& scriptPubKey = mi->second.tx->vout[input.prevout.n].scriptPubKey;
         const CAmount& amount = mi->second.tx->vout[input.prevout.n].nValue;
         SignatureData sigdata;
-        if (!ProduceSignature(*this, TransactionSignatureCreator(&txNewConst, nIn, amount, SIGHASH_ALL), scriptPubKey, sigdata)) {
+        if (!ProduceSignature(*this, MutableTransactionSignatureCreator(&tx, nIn, amount, SIGHASH_ALL), scriptPubKey, sigdata)) {
             return false;
         }
         UpdateTransaction(tx, nIn, sigdata);
@@ -3040,14 +3039,13 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CTransac
 
         if (sign)
         {
-            CTransaction txNewConst(txNew);
             int nIn = 0;
             for (const auto& coin : selected_coins)
             {
                 const CScript& scriptPubKey = coin.txout.scriptPubKey;
                 SignatureData sigdata;
 
-                if (!ProduceSignature(*this, TransactionSignatureCreator(&txNewConst, nIn, coin.txout.nValue, SIGHASH_ALL), scriptPubKey, sigdata))
+                if (!ProduceSignature(*this, MutableTransactionSignatureCreator(&txNew, nIn, coin.txout.nValue, SIGHASH_ALL), scriptPubKey, sigdata))
                 {
                     strFailReason = _("Signing transaction failed");
                     return false;
```

martinus replied:
Thanks! I have now removed the template for `MutableTransactionSignatureCreator`, used your change, ran the format script, and squashed everything

martinus commented:
As suggested by @sipa I have now removed the template for `MutableTransactionSignatureCreator` and used it directly in wallet.cpp.
Also I ran the format script and squashed everything.

sipa commented:
utACK d4ee6530535753292f73a8096a714b1be80f729f

Empact commented on src/script/sign.h:
```diff
@@ -37,26 +37,20 @@ class BaseSignatureCreator {
     int nHashType;
     CAmount amount;
-    const TransactionSignatureChecker checker;
+    const GenericTransactionSignatureChecker<CMutableTransaction> checker;
```
Use can use `MutableTransactionSignatureChecker` here.

maflcko reviewed (approved):
utACK d4ee653053 beside style-nits

MarcoFalke commented:
Tagged with "refactoring" since the only change should be removing the redundant member `MutableTransactionSignatureCreator::tx`

martinus commented:
In 6b8b63a I've fixed all the nits, tried to keep the diff minimal, and squashed it again

MarcoFalke commented:
re-utACK 6b8b63af1461dc11ffd813401e2c36fa44656715

sipa commented:
reutACK 6b8b63af1461dc11ffd813401e2c36fa44656715

fanquake commented:
master (472fe8a):
```
$ time src/test/test_bitcoin -- run_test=transaction_tests/test_big_witness_transaction
Running 1 test case...
*** No errors detected
real	0m20.634s
user	0m21.253s
sys	0m0.163s
```
```
$ time src/test/test_bitcoin
Running 291 test cases...
*** No errors detected
real	1m23.903s
user	0m56.010s
sys	0m40.136s
```

This PR (6b8b63a):
```
time src/test/test_bitcoin --run_test=transaction_tests/test_big_witness_transaction
Running 1 test case...
*** No errors detected
real	0m4.316s
user	0m4.961s
sys	0m0.157s
```

```
time src/test/test_bitcoin
Running 291 test cases...
*** No errors detected
real	1m7.374s
user	0m38.687s
sys	0m41.480s
```

laanwj commented:
utACK 6b8b63af1461dc11ffd813401e2c36fa44656715
